src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -3340,6 +3332,73 @@ private RubyNode addNewlineIfNeeded(ParseNode jrubyNode, RubyNode node) {
         return node;
     }
 
+    private interface ValueFromNode {
+
+        RubyNode prepareAndThen(SourceIndexLength sourceSection, RubyNode subsequent);
+
+        ParseNode get(SourceIndexLength sourceSection);
+
+    }
+
+    private class ValueFromEffectNode implements ValueFromNode {
+
+        private final ParseNode node;
+        private final String temp = environment.allocateLocalTemp("value");
+
+        private boolean sequenced = false;
+
+        public ValueFromEffectNode(ParseNode node) {
+            this.node = node;
+        }
+
+        @Override
+        public RubyNode prepareAndThen(SourceIndexLength sourceSection, RubyNode subsequent) {
+            if (sequenced) {
+                throw new UnsupportedOperationException("don't use a value more than once");
+            }
+            sequenced = true;
+            return sequence(
+                    sourceSection,
+                    Arrays.asList(
+                            new LocalAsgnParseNode(sourceSection, temp, 0, node).accept(BodyTranslator.this),
```

Comment:
It seems much simpler and more efficient to directly build the `WriteLocalVariableNode` here.

Reply:
Can do.

Reply:
Fixed.